Doc/library/__main__.rst

```diff
@@ -1,25 +1,136 @@
-
-:mod:`__main__` --- Top-level script environment
-================================================
+:mod:`__main__` --- CLIs, import-time behavior, and ``__name__ == '__main__'``
+==============================================================================
 
 .. module:: __main__
-   :synopsis: The environment where the top-level script is run.
+   :synopsis: CLIs, import-time behavior, and ``__name__ == '__main__'``
 
 --------------
 
-``'__main__'`` is the name of the scope in which top-level code executes.
-A module's __name__ is set equal to ``'__main__'`` when read from
-standard input, a script, or from an interactive prompt.
+In Python, ``__main__`` is not a single mechanism in the language, but in fact
+is part of two quite different constructs:
+
+1. The ``__name__ == '__main__'`` statement
+2. The ``__main__.py`` file in a Python packages
+
+Each of these mechanisms are related to Python :ref:`tut-modules`; both how
+users interact with them as well as how they interact with each other. See
+:ref:`tut-modules` for details.
+
+
+``__name__ == '__main__'``
+---------------------------
+
+``'__main__'`` is the name of the environment where top-level code is run.
+"Top-level code" means when a Python module is initialized from an interactive
+prompt, from standard input, from a file argument, from a :option:`-c` argument
+or from a :option:`-m` argument, but **not** when it is initialized from an
+import statement.  In any of these situations, the module's ``__name__`` is set
+equal to ``'__main__'``.  The only other context in which Python code is run is
+when it is imported through an import statement. In that case, ``__name__`` is
+set equal to the module's name; usually the name of the file without the
+``.py`` extension.
+
+As a result, a module can discover whether or not it is running in the
+top-level environment by checking its own ``__name__``, which allows a common
+idiom for conditionally executing code when the module is not initialized from
+an import statement::
+
+    if __name__ == '__main__':
+        # Execute when the module is not initialized from an import statement.
+        ...
+
+Design Patterns
+^^^^^^^^^^^^^^^
+
+Putting as few statements as possible in the block below ``if __name___ ==
+'__main__'`` can improve the clarity of your code. Most often, a function named
+``main`` encapuslates the program's "main" behavior, creating this pattern::
+
+    # echo.py
+
+    import sys
+
+    def main(phrase: str):
+       "Print the string to standard output"
+       print(phrase)
+
+    if __name__ == '__main__':
+        main(' '.join(sys.argv))
```

Comment:
I think it's important to use the convention setup by console script entry points

```suggestion
        sys.exit(main())
```

Reply:
As far as I can tell from a grep through `Doc`, this convention is not currently documented anywhere, so I added an explanation of the idiom as well. See 2a398ef21eedff5c88c39df9df599177ba35d952